Userland/Libraries/LibCrypto/Curves/SECPxxxr1.h

```diff
@@ -0,0 +1,700 @@
+        };
+
+        // Convert the input point into Montgomery form
+        point.x = to_montgomery(point.x);
+        point.y = to_montgomery(point.y);
+        point.z = to_montgomery(point.z);
+
+        // Check that the point is on the curve
+        if (!is_point_on_curve(point))
+            return Error::from_string_literal("SECPxxxr1: point is not on the curve");
+
+        JacobianPoint result { 0, 0, 0 };
+        JacobianPoint temp_result { 0, 0, 0 };
+
+        // Calculate the scalar times point multiplication in constant time
+        for (size_t i = 0; i < KEY_BIT_SIZE; i++) {
+            point_add(temp_result, result, point);
+
+            auto condition = (scalar & 1u) == 1u;
+            result.x = select(result.x, temp_result.x, condition);
+            result.y = select(result.y, temp_result.y, condition);
+            result.z = select(result.z, temp_result.z, condition);
+
+            point_double(point, point);
+            scalar >>= 1u;
+        }
+
+        // Convert from Jacobian coordinates back to Affine coordinates
+        convert_jacobian_to_affine(result);
+
+        // Make sure the resulting point is on the curve
+        VERIFY(is_point_on_curve(result));
+
+        // Convert the result back from Montgomery form
+        result.x = from_montgomery(result.x);
+        result.y = from_montgomery(result.y);
+        // Final modular reduction on the coordinates
+        result.x = modular_reduce(result.x);
+        result.y = modular_reduce(result.y);
+
+        // Export the values into an output buffer
+        auto buf = TRY(ByteBuffer::create_uninitialized(POINT_BYTE_SIZE));
+        buf[0] = 0x04;
+        export_big_endian(result.x, buf.bytes().slice(1, KEY_BYTE_SIZE));
+        export_big_endian(result.y, buf.bytes().slice(1 + KEY_BYTE_SIZE, KEY_BYTE_SIZE));
+        return buf;
+    }
+
+    ErrorOr<ByteBuffer> derive_premaster_key(ReadonlyBytes shared_point) override
+    {
+        VERIFY(shared_point.size() == POINT_BYTE_SIZE);
+        VERIFY(shared_point[0] == 0x04);
+
+        ByteBuffer premaster_key = TRY(ByteBuffer::create_uninitialized(KEY_BYTE_SIZE));
+        premaster_key.overwrite(0, shared_point.data() + 1, KEY_BYTE_SIZE);
+        return premaster_key;
+    }
+
+    ErrorOr<bool> verify(ReadonlyBytes hash, ReadonlyBytes pubkey, ReadonlyBytes signature)
+    {
+        Crypto::ASN1::Decoder asn1_decoder(signature);
+        TRY(asn1_decoder.enter());
+
+        auto r_bigint = TRY(asn1_decoder.read<Crypto::UnsignedBigInteger>(Crypto::ASN1::Class::Universal, Crypto::ASN1::Kind::Integer));
+        auto s_bigint = TRY(asn1_decoder.read<Crypto::UnsignedBigInteger>(Crypto::ASN1::Class::Universal, Crypto::ASN1::Kind::Integer));
+
+        StorageType r = 0u;
+        StorageType s = 0u;
+        for (size_t i = 0; i < (KEY_BIT_SIZE / 32); i++) {
+            StorageType rr = r_bigint.words()[i];
+            StorageType ss = s_bigint.words()[i];
+            r |= (rr << (i * 32));
```

Comment:
I don't think this will work, as `r_bigint.words()` is a reference to a `Vector`, not an actual array.

Reply:
Vector has `::data`

Reply:
Oh, this would still require a rather ugly `*reinterpret_cast<u32(*)[KEY_BIT_SIZE / 32]>(r_bigint.words().data())`. So, feel free to not do anything with this.